### PR TITLE
Rule S3440: add support for C# 9 pattern matching

### PR DESCRIPTION
**Pull request by dmitry-golovinov-sonarsource**

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs [hunk 1]
```diff
@@ -3,12 +3,49 @@
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs:3]: Please try to cover or at least document the following FNs:
```
object a = null;

switch (a)
{
    case null:
        {
            a = null;
        }
        break;
    default:
        break;
}

if ((x,y) is (1,2))
{
    x = 1;
    y = 2;
}
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignment.cs [hunk 2]
```diff
@@ -19,13 +19,15 @@
  */
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignment.cs:19]: I see that there is a codefix  provider for the rule. Lets try to update that too to work with c#9 patterns if it is a low hanging fruit.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
The changes made are good, and lets try to cover some more cases.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs (hunk 1 above)
Comment (dmitry-golovinov-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs:3]: Documented them as FNs for now

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignment.cs [hunk 3]
```diff
@@ -36,58 +38,80 @@ public sealed class RedundantConditionalAroundAssignment : SonarDiagnosticAnalyz
...
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, condition.GetLocation()));
+            }
+        }
+
+        private static void UselessConditionSwitchExpression(SyntaxNodeAnalysisContext c)
+        {
+            var switchExpression = (SwitchExpressionSyntaxWrapper)c.Node;
+
+            if (switchExpression.Arms.Count != 1
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignment.cs:85]: Because of this check we do not cover something like:
```
int y = 6;

y = y switch
{
    5 => 5, // Noncompliant
    6 => 6 // Noncompliant
};
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.Fixed.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
+﻿int x = 5;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.Fixed.cs:1]: A case that we could also cover is the following:
```
y = y switch
{
    _ => y
};
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignmentCodeFixProvider.cs [hunk 5]
```diff
@@ -46,11 +47,24 @@ protected override Task RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext c
             var condition = root.FindNode(diagnosticSpan) as ExpressionSyntax;
             var ifStatement = condition?.FirstAncestorOrSelf<IfStatementSyntax>();
 
-            if (ifStatement == null)
+            if (ifStatement != null)
+            {
+                return HandleIfStatement(root, context, ifStatement);
+            }
+
+            var switchExpression = root.FindNode(diagnosticSpan).FirstAncestorOrSelf((SyntaxNode x) => x.IsKind(SyntaxKindEx.SwitchExpression));
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignmentCodeFixProvider.cs:55]: This is a bit more readable for me, but feel free to leave it as is if you don`t agree with it.
```suggestion
            var switchExpression = root.FindNode(diagnosticSpan).FirstAncestorOrSelf<SyntaxNode>(x => x.IsKind(SyntaxKindEx.SwitchExpression));
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs [hunk 6]
```diff
@@ -3,14 +3,57 @@
 
 y = y switch
 {
-    not 5 => 5 // Compliant - FN
+    not 5 => 5 // Noncompliant
 };
 
 y = y switch
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs:9]: I would say this is an edge case.  Because y will be set to 5 only if it was 5, otherwise an exception will be thrown.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Last round of additional cases.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs (hunk 6 above)
Comment (dmitry-golovinov-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantConditionalAroundAssignment.CSharp9.cs:9]: No action

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignment.cs [hunk 7]
```diff
@@ -36,58 +38,82 @@ public sealed class RedundantConditionalAroundAssignment : SonarDiagnosticAnalyz
...
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, condition.GetLocation()));
+            }
+        }
+
+        private static void UselessConditionSwitchExpression(SyntaxNodeAnalysisContext c)
+        {
+            var switchExpression = (SwitchExpressionSyntaxWrapper)c.Node;
+
+            if (!(switchExpression.SyntaxNode.GetFirstNonParenthesizedParent() is AssignmentExpressionSyntax parentExpression)
+                || !CSharpEquivalenceChecker.AreEquivalent(parentExpression.Left, switchExpression.GoverningExpression))
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignment.cs:86]: Do we need this equivalence check? I think because of it we do not handle cases like:
```
y = x switch
{
    _ => 5       // Noncompliant
};
```

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Great work, once last case left to cover.

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!
